docs/deploying_an_application.md

```diff
@@ -0,0 +1,123 @@
+
+## Procfile
+
+Empire was modeled after the Heroku API.  As such, many concepts and commands are similar.  One concept which has bled over and which is a requirement is that of a [Procfile][procfiles].  For your application to actually work, you must include a Procfile at the root level of your application.
+
+Imagine a Python application which simply ran celery.  A `Procfile` for such a service could look like this:
+
+```
+workerservice: celery -A tasks worker --loglevel=info 
+```
+
+The file would be named `Procfile`, and live at the directory root for your application.
+
+```console
+$ tree .
+.
+├── Dockerfile
+├── Procfile
+├── myapplication/ 
+
+```
+
+
+## Empire application types
+
+Empire treats "web" and "non-web" services different.  Here, a "web" process is defined as anything which needs to expose a port.  The way to differentiate the two types of processes is easy.  
+
+### Non-web processes
+
+Let's start with non-web processes first since they're much simpler.  If your service doesn't expose a port it doesn't need to be discovered by any other system.  With that, all you need to do it deploy your application and let Empire manage it. The act of discovering *other* services is up to you (database, caches, etc.)
+
+### Web processes
+
+If your service is going to be used by other systems, you'll need to run some type of server which exposes a port.  In this scenario you'll need to name it `web` in your Procfile:
+
+```
+web: node server.js
+```
+
+or, using Django as an example:
+
+```
+web: python manage.py runserver
+```
+
+For web applications, Empire does a few things.  There are multiple layers of routing, none of which are extremely complex.
+
+Assume we have deployed a container named `mycompany/awesome-app`.
+
+The routing to your application is handled as such inside of the VPC:
+
+```
+
+     http://awesome-app/  
+              +           
+              |           
+              |           
+              v           
+             ELB: port 80
+              +           
+              |           
+              |           
+              v           
+            Minion: port [9000-10000]
+              +           
+              |           
+              |           
+              v           
+          Container: port $PORT
+
+```
+
+There are various things going on...let's break them down:
+
+- Empire creates an internal HostedZone CNAME record for `awesome-app`.  This CNAME points at an ELB which Empire also creates, specifically for this application.  If you `emp deploy mycompany/another-app`, yet *another* CNAME and ELB would be created for `another-app`.
+- The ELB created is managed by Empire. As you scale your application *up*, instances running your application are placed into the ELB.  Likewise, as you scale your application *down*, instances are removed from the ELB. The ELB listens on port 80 and maps to a random port between 9000 and 10000 on the minion instances running your application.
```

Comment:
hmm - random isn't quite right, though it's probably safe enough for this.  Might be worth mentioning that each app gets a port assigned to it - so you won't have two apps with the same port.  We actually just pick the next port avaiable, incrementing from 9000-10000.  Not sure of an easy way to explain that.